vowpalwabbit/config/options_cli.cc

```diff
@@ -0,0 +1,486 @@
+  for (auto& token : current_tokens) { result_tokens.push_back(token); }
+}
+
+void consume_short_option(const std::map<char, std::shared_ptr<base_option>>& known_short_options,
+    /*in-out*/ std::queue<VW::string_view>& command_line,
+    /*out*/ std::unordered_map<VW::string_view, std::vector<VW::string_view>>& result)
+{
+  auto current_token = command_line.front();
+
+  // Check if just dangling -
+  if (current_token.size() == 1)
+  {
+    const auto result_name = "__POSITIONAL__";
+    result[result_name].push_back(current_token);
+    command_line.pop();
+    return;
+  }
+
+  auto current_opt = current_token[1];
+
+  std::vector<VW::string_view> current_tokens;
+  if (current_token.size() > 2)
+  {
+    auto opt_value = current_token.substr(2);
+    current_tokens.push_back(opt_value);
+  }
+
+  // Consume this token.
+  command_line.pop();
+
+  // See if there is a known option for this name
+  auto known_opt = known_short_options.find(current_opt);
+  if (known_opt == known_short_options.end())
+  {
+    // If there is no known option we are going to just treat it as an unknown option.
+    result["__POSITIONAL__"].push_back(current_token);
+    return;
+  }
+
+  const auto& option_spec = *known_opt->second;
+  consume_tokens(option_spec, command_line, current_tokens);
+
+  // Will create vector if not exists - this is intentional.
+  auto& result_tokens = result[option_spec.m_name];
+  for (auto& token : current_tokens) { result_tokens.push_back(token); }
+}
+
+namespace VW
+{
+namespace config
+{
+namespace details
+{
+struct cli_typed_option_handler : typed_option_visitor
+{
+  options_cli& m_options;
+  std::unordered_map<VW::string_view, std::vector<VW::string_view>>& m_tokens;
+
+  cli_typed_option_handler(
+      options_cli& options, std::unordered_map<VW::string_view, std::vector<VW::string_view>>& tokens)
+      : m_options(options), m_tokens(tokens)
+  {
+  }
+
+  void visit(typed_option<uint32_t>& option) override { handle_typed_option(option); }
+  void visit(typed_option<uint64_t>& option) override { handle_typed_option(option); }
+  void visit(typed_option<int64_t>& option) override { handle_typed_option(option); }
+  void visit(typed_option<int32_t>& option) override { handle_typed_option(option); }
+  void visit(typed_option<bool>& option) override { handle_typed_option(option); }
+  void visit(typed_option<float>& option) override { handle_typed_option(option); }
+  void visit(typed_option<std::string>& option) override { handle_typed_option(option); }
+  void visit(typed_option<std::vector<std::string>>& option) override { handle_typed_option(option); }
+
+  template <typename T>
+  void handle_typed_option(typed_option<T>& option)
+  {
+    auto tokens_it = m_tokens.find(option.m_name);
+    bool option_was_supplied = tokens_it != m_tokens.end();
+
+    // Default case
+    if (!option_was_supplied)
+    {
+      if (option.default_value_supplied()) { option.value(option.default_value(), true); }

```

Comment:
what happens to this option if it was not supplied and there is no default?

Reply:
No value is written, it is essentially untouched